### ISSUE TITLE
slow basic FS operations on large git-annex repositories .git/annex'es due to KEY/ directories

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#32]
echoing discussion we are having in #17, I wanted to also bring one about non-direct (i.e. regular/classical) annex'es and their impact.  As http://git-annex.branchable.com/internals/ and http://git-annex.branchable.com/internals/lockdown/ outlines, to prevent accidental removal of the file, every file is placed under `.git/annex/objects/aa/bb/KEY/KEY` directory. where for KEY directory 'w' permission is taken away so the KEY file can't be removed.  It indeed solves the accidental removal issue, but
1. it nearly doubles inodes needed for the repository.  So on large ones, impact on the file system might be notable (on smaug I have 100mil inodes in use now :-/ could have been just 50mil? ;))
2. operations (even a simple du or ls -R) might take MUCH longer since leaf directories are to contain a single file and thus there is quite a heavy file hierarchy traversal is needed.

There was a recent comment http://git-annex.branchable.com/internals/lockdown/#comment-f77526824d026f213ea98939fda9ac4c possibly on linux specific way, but I wondered:  Why just not to take 'w' permission away from `.git/annex/objects/aa/bb` directories, and store files under.  Yes -- for any get/drop that directory permission would need to flip to being writable but with proper guards around, and fsck checking, (or just adding some lock file which if not removed would signal that one of the underlying directories might have been left writable) I bet it should be feasible to make it work quite reliably without sacrificing FS meta-information real estate/performance?

**Comment by joeyh** [comment]
Yaroslav Halchenko wrote:
>  2. operations (even a simple du or ls -R) might take MUCH longer since leaf
>     directories are to contain a single file and thus there is quite a heavy
>     file hierarchy traversal is needed.

I am doubtful about this without seeing some measurement.

The full traversal is .git/annex/xxx/yyy/KEY/obj, possibly with
../../ in front. Does removing one item from this path really impact
much? Note that the yyy directory is rarely going to be in cache either.

Also, any change in this area would break compatability with every
existing git-annex repository in a horrible way, as the path is
committed to git. Rewriting all the paths would bloat the git repository
size, and leave files inaccessible when checking out old tags etc.

-- 
see shy jo

**Comment by yarikoptic** [comment]
indeed, yyy would unlikely to be in the cache ... but it is not just about the cache  -- next thought was if it was xxx/yy (or even xx/yy) then collisions in large repositories would lead to growing number of files under yy, while still allowing for efficient lookup and also making reading that directory content more efficient due to having multiple keys stored at that level.
yeah -- this food for though is scary and might be just be highlighting the fact that a good middle ground could be hard to achieve throughout all the use cases and asking for supporting multiple incompatible layouts... but that is why I am experimenting/checking now before actually creating a bunch of annexes to share ;)
anyways -- I will try to get some timing numbers to argue without empty hands -- it could well be that pros don't outweight cons ;-)

**Comment by yarikoptic** [comment]
just preliminary results for now.  only on ZFS with any ARC (1st or 2nd level) disabled since it is impossible afaik just to flush those read caches.  -xx-yy-key is the original git annex layout  and -xx-yy is where I moved files from their per-file directory.  annex keeps around 12,000 files .  Such a directory is one of the 500 I have and even managed to merge into a single annex I was talking about before ;)

```
*$> ./timeit.sh 992774-xx-yy-key 992774-xx-yy
= Command: du -scm 992774-xx-yy-key
== Cold du -scm 992774-xx-yy-key
1482    992774-xx-yy-key
1482    total
        1:54.20 real,   0.60 user,      22.03 sys
== Warm/repeated
1482    992774-xx-yy-key
1482    total
        1:19.53 real,   0.44 user,      10.40 sys
= Command: du -scm 992774-xx-yy
== Cold du -scm 992774-xx-yy
724     992774-xx-yy
724     total
        1:04.68 real,   0.32 user,      15.70 sys
== Warm/repeated
724     992774-xx-yy
724     total
        0:39.24 real,   0.18 user,      5.54 sys


= Command: tar -cf 992774-xx-yy-key.tgz 992774-xx-yy-key
== Cold tar -cf 992774-xx-yy-key.tgz 992774-xx-yy-key
        2:18.79 real,   3.23 user,      28.32 sys
== Warm/repeated
        1:31.54 real,   2.45 user,      13.09 sys
= Command: tar -cf 992774-xx-yy.tgz 992774-xx-yy
== Cold tar -cf 992774-xx-yy.tgz 992774-xx-yy
        1:34.76 real,   1.94 user,      17.86 sys
== Warm/repeated
        1:08.95 real,   1.56 user,      7.84 sys


= Command: chmod +w -R 992774-xx-yy-key/.git/annex/objects
== Cold chmod +w -R 992774-xx-yy-key/.git/annex/objects
        1:52.10 real,   0.44 user,      19.15 sys
== Warm/repeated
        1:19.14 real,   0.32 user,      10.68 sys
= Command: chmod +w -R 992774-xx-yy/.git/annex/objects
== Cold chmod +w -R 992774-xx-yy/.git/annex/objects
        0:58.14 real,   0.27 user,      10.69 sys
== Warm/repeated
        0:39.73 real,   0.21 user,      5.22 sys


= Command: rm -rf 992774-xx-yy-key
== Cold rm -rf 992774-xx-yy-key
        3:25.58 real,   0.54 user,      40.51 sys
= Command: rm -rf 992774-xx-yy
== Cold rm -rf 992774-xx-yy
        1:41.23 real,   0.34 user,      18.96 sys
```
As expected, all run times are cut by 1/3, besides that rm seems to be cut more than in half.  These timings were from a single run only.  Also amount of file system real estate, at least with this obscure test where files are actually tiny is cut nearly in half.  I expect also that on larger repos, with more collisions under xx/yy directories difference in performance will be even larger.  I might run such a test later on.
If interested, original location was smaug:/mnt/scrap/datalad/mimic_repos/hcp-ls.20141020-500subjects-addurl  and all the scripts used to produce those two directories and run these benchmarks are now available from https://github.com/datalad/datalad/tree/master/tools/benchmarking . I will generalize it all into a single command to be ran later on multiple locations

**Comment by yarikoptic** [comment]
ok -- did a bit more of silly benchmarking which also seems have pointed to two weak RAM modules in the new box ;)  I have added also -xx-y  layout where for the 2nd level only one character is used, and then -deref -- as if data was never annex'ed (present only in the last results for 120,000 files): http://www.onerussian.com/tmp/benchs.txt .  So thing are pretty much boiling down to twice speed up on large repos from going away from xx/yy/key to xx/yy, and then also twice to allowing more collisions (of which there weren't way too many -- up to 9 -- with 120,000 files) via using just xx/y.   Since I have no plans for that box for this night -- let's make it sweat with 1,200,000 files... ;)

**Comment by yarikoptic** [comment]
@joeyh  -- here is that ultimate run for 1,200,000 files (above url on my website also should contain it).  Using an xx/y/ layout with more collisions (but not much really -- just goes up to few with 62 collisions) provides notable advantages in terms of size and performance impact

```
I: stats
test1200000-xx-yy-key:  size: 154091M   inodes: 8961913 links: 1200000  2nd level collisions: 1:381573 10:1 2:218490 3:83650 4:23750 5:5526 6:1078 7:163 8:22 9:8 
test1200000-xx-yy:      size: 35856M    inodes: 3116333 links: 1200000  2nd level collisions: 1:381573 10:1 2:218490 3:83650 4:23750 5:5526 6:1078 7:163 8:22 9:8 
test1200000-xx-y:       size: 4074M     inodes: 2434840 links: 1200000  2nd level collisions: 13:2 16:1 17:3 18:9 19:16 20:26 21:55 22:102 23:155 24:216 25:323 26:511 27:623 28:865 29:961 30:1300 31:1554 32:1620 33:1846 34:2029 35:2146 36:2123 37:2076 38:2101 39:1898 40:1813 41:1549 42:1381 43:1216 44:961 45:795 46:675 47:466 48:373 49:277 50:204 51:154 52:102 53:79 54:52 55:42 56:25 57:18 58:11 59:9 61:1 62:4 
test1200000-deref:      size: 618M      inodes: 1201002 links: 0    2nd level collisions: NA


= Command: du -sm test1200000-deref 
== Cold: 618 test1200000-deref
    0:52.29 real,   0.67 user,  26.85 sys
== Warm: 618    test1200000-deref
    0:10.64 real,   0.37 user,  4.30 sys
= Command: du -sm test1200000-xx-y 
== Cold: 4074   test1200000-xx-y
    1:46.54 real,   1.88 user,  67.19 sys
== Warm: 4074   test1200000-xx-y
    0:25.23 real,   0.88 user,  11.00 sys
= Command: du -sm test1200000-xx-yy 
== Cold: 35856  test1200000-xx-yy
    10:25.69 real,  7.57 user,  254.21 sys
== Warm: 35856  test1200000-xx-yy
    8:17.42 real,   8.26 user,  191.94 sys
= Command: du -sm test1200000-xx-yy-key 
== Cold: 154091  test1200000-xx-yy-key
    28:52.08 real,  29.51 user, 613.93 sys
== Warm: 154091 test1200000-xx-yy-key
    17:24.76 real,  21.01 user, 296.61 sys


= Command: tar -cf test1200000-deref.tgz test1200000-deref 
== Cold:    1:39.41 real,   3.86 user,  62.84 sys
== Warm:    0:45.35 real,   2.83 user,  33.19 sys
= Command: tar -cf test1200000-xx-y.tgz test1200000-xx-y 
== Cold:    5:16.26 real,   62.94 user, 174.10 sys
== Warm:    3:18.12 real,   61.89 user, 108.01 sys
= Command: tar -cf test1200000-xx-yy.tgz test1200000-xx-yy 
== Cold:    13:57.58 real,  93.17 user, 358.36 sys
== Warm:    10:12.85 real,  80.26 user, 232.72 sys
= Command: tar -cf test1200000-xx-yy-key.tgz test1200000-xx-yy-key 
== Cold:    1:43:01 real,   257.89 user,    1750.94 sys
== Warm:    40:53.03 real,  158.88 user,    948.32 sys


= Command: chmod +w -R test1200000-xx-y/.git/annex/objects 
== Cold:    1:43.88 real,   1.23 user,  62.97 sys
== Warm:    0:17.84 real,   0.24 user,  17.19 sys
= Command: chmod +w -R test1200000-xx-yy/.git/annex/objects 
== Cold:    9:53.64 real,   7.98 user,  259.94 sys
== Warm:    8:25.23 real,   6.14 user,  213.53 sys
= Command: chmod +w -R test1200000-xx-yy-key/.git/annex/objects 
== Cold:    22:19.55 real,  17.09 user, 501.07 sys
== Warm:    18:10.13 real,  14.58 user, 349.21 sys


= Command: rm -rf test1200000-deref 
== Cold:    1:37.02 real,   0.65 user,  81.36 sys
= Command: rm -rf test1200000-xx-y 
== Cold:    4:05.82 real,   1.99 user,  207.38 sys
= Command: rm -rf test1200000-xx-yy 
== Cold:    14:09.17 real,  10.73 user, 451.76 sys
= Command: rm -rf test1200000-xx-yy-key 
== Cold:    44:52.93 real,  31.27 user, 1274.58 sys
```

**Comment by joeyh** [comment]
I have made a page  that is summarising these proposed changes and their
cost/benefit tradeoffs.

There is also a proposal at the end of the page that I'd appreciate
comments on, under "living on the edge".

-- 
see shy jo

**Comment by yarikoptic** [comment]
@joeyh url?

**Comment by joeyh** [comment]
Meant to say it's http://git-annex.branchable.com/design/v6

-- 
see shy jo

**Comment by yarikoptic** [comment]
Thank you @joeyh for giving it all such a thorough consideration!  note that the speed-up wasn't specific to 'spindles' and also present for SSD discs:

```
*** smaug: ext4 on a mirrored SSD (samsungs)
test12000-xx-yy-key:     links: 12000    2nd level collisions: 1:11856 2:72
test12000-xx-yy:         links: 12000    2nd level collisions: 1:11856 2:72
test12000-xx-y:  		 links: 12000    2nd level collisions: 1:8307  2:1520 3:193 4:17 6:1

= Command: du -sm test12000-xx-y
== Cold: [sudo] password for yoh:
301     test12000-xx-y
        0:01.82 real,   0.05 user,      0.56 sys
== Warm: 301    test12000-xx-y
        0:00.13 real,   0.02 user,      0.10 sys
= Command: du -sm test12000-xx-yy
== Cold: 308    test12000-xx-yy
        0:02.01 real,   0.07 user,      0.52 sys
== Warm: 308    test12000-xx-yy
        0:00.13 real,   0.04 user,      0.09 sys
= Command: du -sm test12000-xx-yy-key
== Cold: 355    test12000-xx-yy-key
        0:03.69 real,   0.09 user,      0.92 sys
== Warm: 355    test12000-xx-yy-key
        0:00.27 real,   0.04 user,      0.22 sys


= Command: tar -cf test12000-xx-y.tgz test12000-xx-y
== Cold:        0:11.34 real,   0.79 user,      2.91 sys
== Warm:        0:00.82 real,   0.30 user,      0.48 sys
= Command: tar -cf test12000-xx-yy.tgz test12000-xx-yy
== Cold:        0:11.30 real,   1.03 user,      2.84 sys
== Warm:        0:01.15 real,   0.36 user,      0.66 sys
= Command: tar -cf test12000-xx-yy-key.tgz test12000-xx-yy-key
== Cold:        0:15.51 real,   1.62 user,      4.30 sys
== Warm:        0:01.31 real,   0.57 user,      0.63 sys


= Command: chmod +w -R test12000-xx-y/.git/annex/objects
== Cold:        0:01.51 real,   0.04 user,      0.42 sys
== Warm:        0:00.12 real,   0.02 user,      0.09 sys
= Command: chmod +w -R test12000-xx-yy/.git/annex/objects
== Cold:        0:01.79 real,   0.05 user,      0.52 sys
== Warm:        0:00.17 real,   0.02 user,      0.15 sys
= Command: chmod +w -R test12000-xx-yy-key/.git/annex/objects
== Cold:        0:03.66 real,   0.12 user,      0.94 sys
== Warm:        0:00.21 real,   0.04 user,      0.16 sys


= Command: rm -rf test12000-xx-y
== Cold:        0:02.78 real,   0.10 user,      1.48 sys
= Command: rm -rf test12000-xx-yy
== Cold:        0:03.29 real,   0.05 user,      1.88 sys
= Command: rm -rf test12000-xx-yy-key
== Cold:        0:05.36 real,   0.14 user,      2.59 sys
*** smaug: ZFS (ARCs L1 and L2 are on, and for all, not just metadata)
test12000-xx-yy-key:     links: 12000    2nd level collisions: 1:11856 2:72
test12000-xx-yy:         links: 12000    2nd level collisions: 1:11856 2:72
test12000-xx-y:          links: 12000    2nd level collisions: 1:8307 2:1520 3:193 4:17 6:1
```

**Comment by joeyh** [comment]
I have now implemented a framework for tuning git-annex repositories.
See http://git-annex.branchable.com/tuning/
Needs git-annex commit b0575c621f46fcc876fab4132dede6343c68c3ef or newer.

Tuning a repository to use annex.tune.branchhash1 may cut down on the
size of the git-annex branch, and also on the speed of lookups, as it
eliminates one level of the hash directories. This is as discussed in
issue #17.

Tuning a repository to use annex.tune.objecthash1 avoids one level of
hash directories under .git/annex/objects. This should improve performance
per issue #32.

Both these tunings are pretty basic; it's easier to drop a single hash
directory than to make larger changes to layout. However, it should be
easy to add more tunings along these lines now that the groundwork is laid.

Probably what it would make sense for Yaroslav/somebody to do now is build
git-annex HEAD and use it to test rebuilds of the large repo (or a
synthetic test repo) with both tunings, and see how much of a win they are.

-- 
see shy jo

**Comment by yarikoptic** [comment]
thanks @joeyh -- building now and will test/compare tomorrow on few cases. 

- when do you see objecthashlower=true  being beneficial?  it would result in a smaller number of subdirectories, so I guess for even smaller repos it might be of some benefit but still not clear to me when should I use it?

- theoretically for small/regular repos logical to have objecthash1=true a default, and then increase if repository is expected to be so big that 1 level of hashing might result in performance hit, right?

- neither of those 3 takes care about removing the "ro" protection directory layer (which is what was of the major impact), right?

**Comment by joeyh** [comment]
Yaroslav Halchenko wrote:
>   • when do you see objecthashlower=true being beneficial? it would result in a
>     smaller number of subdirectories, so I guess for even smaller repos it
>     might be of some benefit but still not clear to me when should I use it?

I didn't add that one for datalad but just because it's a common
complaint that git-annex uses mixed case hash directories.

>   • theoretically for small/regular repos logical to have objecthash1=true a
>     default, and then increase if repository is expected to be so big that 1
>     level of hashing might result in performance hit, right?

Theoretically, but there is currently no way to transition between these
sorts of repositories.

>   • neither of those 3 takes care about removing the "ro" protection directory
>     layer (which is what was of the major impact), right?

Removing one level of hash subdirectories should have similar affects on
FS caching.

-- 
see shy jo

**Comment by yarikoptic** [comment]
>> • neither of those 3 takes care about removing the "ro" protection directory layer (which is what was of the major impact), right?
> Removing one level of hash subdirectories should have similar affects on FS caching.

Here is the gory results of me running tests (+ cloning, annex get, annex drop) on previous layouts + 3 new ones: -bh1 for branchhash1, -oh1 for objecthash1, -oh1+bh1 for both of those at once, and -ohl for objecthashlower. Unfortunately output is not easy to digest but few points stand: objecthash1 allows indeed to cut some time and storage but far not as much as just dropping those ro directories.  And I haven't tried yet -xx, which would be the objecthash1 with dropped ro directories

./test1200x10-deref           : size: 48M	inodes: 12012	links: 0	2nd level collisions: NA
./test1200x10-oh1             : size: 336M	inodes: 83792	links: 12000	2nd level collisions: 1:12000 
./test1200x10-ohl             : size: 394M	inodes: 98638	links: 12000	2nd level collisions: 1:11992 2:4 
./test1200x10-bh1             : size: 334M	inodes: 83720	links: 12000	2nd level collisions: 1:11856 2:72 
./test1200x10-oh1+bh1         : size: 334M	inodes: 83720	links: 12000	2nd level collisions: 1:11856 2:72 
./test1200x10-x-xy            : size: 147M	inodes: 35152	links: 12000	2nd level collisions: 1:8307 2:1520 3:193 4:17 6:1 
./test1200x10-xx-y            : size: 147M	inodes: 35120	links: 12000	2nd level collisions: 1:8307 2:1520 3:193 4:17 6:1 
./test1200x10-xx-yy           : size: 154M	inodes: 37010	links: 12000	2nd level collisions: 1:11856 2:72 
./test1200x10-xx-yy-key       : size: 382M	inodes: 95719	links: 12000	2nd level collisions: 1:11856 2:72 

= Command: du -sm test1200x10-bh1 
== Cold: 334	test1200x10-bh1
	0:03.60 real,	0.12 user,	1.00 sys
== Warm: 334	test1200x10-bh1
	0:00.27 real,	0.05 user,	0.22 sys
= Command: du -sm test1200x10-deref 
== Cold: 48	test1200x10-deref
	0:00.09 real,	0.00 user,	0.07 sys
== Warm: 48	test1200x10-deref
	0:00.02 real,	0.01 user,	0.01 sys
= Command: du -sm test1200x10-oh1 
== Cold: 336	test1200x10-oh1
	0:02.12 real,	0.08 user,	0.66 sys
== Warm: 336	test1200x10-oh1
	0:00.20 real,	0.04 user,	0.16 sys
= Command: du -sm test1200x10-oh1+bh1 
== Cold: 334	test1200x10-oh1+bh1
	0:03.57 real,	0.11 user,	0.99 sys
== Warm: 334	test1200x10-oh1+bh1
	0:00.26 real,	0.05 user,	0.20 sys
= Command: du -sm test1200x10-ohl 
== Cold: 394	test1200x10-ohl
	0:04.54 real,	0.20 user,	1.36 sys
== Warm: 394	test1200x10-ohl
	0:00.36 real,	0.07 user,	0.28 sys
= Command: du -sm test1200x10-x-xy 
== Cold: 147	test1200x10-x-xy
	0:01.83 real,	0.08 user,	0.53 sys
== Warm: 147	test1200x10-x-xy
	0:00.14 real,	0.02 user,	0.11 sys
= Command: du -sm test1200x10-xx-y 
== Cold: 147	test1200x10-xx-y
	0:01.71 real,	0.10 user,	0.44 sys
== Warm: 147	test1200x10-xx-y
	0:00.10 real,	0.02 user,	0.07 sys
= Command: du -sm test1200x10-xx-yy 
== Cold: 154	test1200x10-xx-yy
	0:02.10 real,	0.07 user,	0.64 sys
== Warm: 154	test1200x10-xx-yy
	0:00.16 real,	0.04 user,	0.11 sys
= Command: du -sm test1200x10-xx-yy-key 
== Cold: 382	test1200x10-xx-yy-key
	0:03.60 real,	0.09 user,	0.98 sys
== Warm: 382	test1200x10-xx-yy-key
	0:00.29 real,	0.06 user,	0.22 sys


= Command: tar -cf test1200x10-bh1.tar test1200x10-bh1 
== Cold: 	0:13.00 real,	1.33 user,	3.34 sys
== Warm: 	0:01.48 real,	0.56 user,	0.70 sys
= Command: tar -cf test1200x10-deref.tar test1200x10-deref 
== Cold: 	0:01.55 real,	0.03 user,	0.37 sys
== Warm: 	0:00.11 real,	0.04 user,	0.06 sys
= Command: tar -cf test1200x10-oh1.tar test1200x10-oh1 
== Cold: 	0:12.64 real,	1.08 user,	3.28 sys
== Warm: 	0:01.41 real,	0.52 user,	0.78 sys
= Command: tar -cf test1200x10-oh1+bh1.tar test1200x10-oh1+bh1 
== Cold: 	0:12.85 real,	1.25 user,	3.34 sys
== Warm: 	0:01.46 real,	0.59 user,	0.72 sys
= Command: tar -cf test1200x10-ohl.tar test1200x10-ohl 
== Cold: 	0:14.68 real,	1.44 user,	3.76 sys
== Warm: 	0:01.55 real,	0.62 user,	0.77 sys
= Command: tar -cf test1200x10-x-xy.tar test1200x10-x-xy 
== Cold: 	0:05.87 real,	0.76 user,	1.82 sys
== Warm: 	0:00.91 real,	0.39 user,	0.40 sys
= Command: tar -cf test1200x10-xx-y.tar test1200x10-xx-y 
== Cold: 	0:07.36 real,	1.21 user,	2.20 sys
== Warm: 	0:00.78 real,	0.36 user,	0.41 sys
= Command: tar -cf test1200x10-xx-yy.tar test1200x10-xx-yy 
== Cold: 	0:05.82 real,	0.84 user,	1.88 sys
== Warm: 	0:00.84 real,	0.35 user,	0.47 sys
= Command: tar -cf test1200x10-xx-yy-key.tar test1200x10-xx-yy-key 
== Cold: 	0:15.10 real,	1.60 user,	3.81 sys
== Warm: 	0:01.58 real,	0.53 user,	0.77 sys


= Command: chmod +w -R test1200x10-bh1/.git/annex/objects 
== Cold: 	0:03.45 real,	0.08 user,	0.95 sys
== Warm: 	0:00.29 real,	0.02 user,	0.26 sys
= Command: chmod +w -R test1200x10-oh1/.git/annex/objects 
== Cold: 	0:01.82 real,	0.02 user,	0.51 sys
== Warm: 	0:00.17 real,	0.00 user,	0.16 sys
= Command: chmod +w -R test1200x10-oh1+bh1/.git/annex/objects 
== Cold: 	0:03.81 real,	0.13 user,	1.02 sys
== Warm: 	0:00.28 real,	0.03 user,	0.24 sys
= Command: chmod +w -R test1200x10-ohl/.git/annex/objects 
== Cold: 	0:03.51 real,	0.09 user,	0.90 sys
== Warm: 	0:00.28 real,	0.05 user,	0.22 sys
= Command: chmod +w -R test1200x10-x-xy/.git/annex/objects 
== Cold: 	0:01.58 real,	0.02 user,	0.49 sys
== Warm: 	0:00.13 real,	0.00 user,	0.12 sys
= Command: chmod +w -R test1200x10-xx-y/.git/annex/objects 
== Cold: 	0:01.54 real,	0.04 user,	0.41 sys
== Warm: 	0:00.17 real,	0.02 user,	0.14 sys
= Command: chmod +w -R test1200x10-xx-yy/.git/annex/objects 
== Cold: 	0:01.84 real,	0.06 user,	0.54 sys
== Warm: 	0:00.18 real,	0.02 user,	0.15 sys
= Command: chmod +w -R test1200x10-xx-yy-key/.git/annex/objects 
== Cold: 	0:03.41 real,	0.06 user,	0.96 sys
== Warm: 	0:00.26 real,	0.04 user,	0.22 sys




= Command: bash -c cd test1200x10-bh1.clone; git annex get . 1>/dev/null 2>&1 || :  
== Cold: 	0:00.19 real,	0.04 user,	0.02 sys
= Command: bash -c cd test1200x10-oh1.clone; git annex get . 1>/dev/null 2>&1 || :  
== Cold: 	0:00.17 real,	0.01 user,	0.01 sys
= Command: bash -c cd test1200x10-oh1+bh1.clone; git annex get . 1>/dev/null 2>&1 || :  
== Cold: 	0:00.17 real,	0.01 user,	0.02 sys
= Command: bash -c cd test1200x10-ohl.clone; git annex get . 1>/dev/null 2>&1 || :  
== Cold: 	0:00.17 real,	0.01 user,	0.02 sys
= Command: bash -c cd test1200x10-xx-yy-key.clone; git annex get . 1>/dev/null 2>&1 || :  
== Cold: 	2:38.86 real,	76.63 user,	17.92 sys


= Command: bash -c cd test1200x10-bh1.clone; git annex drop . 1>/dev/null 2>&1 || :  
== Cold: 	0:00.17 real,	0.02 user,	0.03 sys
= Command: bash -c cd test1200x10-oh1.clone; git annex drop . 1>/dev/null 2>&1 || :  
== Cold: 	0:00.16 real,	0.02 user,	0.03 sys
= Command: bash -c cd test1200x10-oh1+bh1.clone; git annex drop . 1>/dev/null 2>&1 || :  
== Cold: 	0:00.16 real,	0.02 user,	0.02 sys
= Command: bash -c cd test1200x10-ohl.clone; git annex drop . 1>/dev/null 2>&1 || :  
== Cold: 	0:00.17 real,	0.01 user,	0.03 sys
= Command: bash -c cd test1200x10-x-xy.clone; git annex drop . 1>/dev/null 2>&1 || :  
== Cold: 	0:35.81 real,	16.60 user,	1.70 sys
= Command: bash -c cd test1200x10-xx-y.clone; git annex drop . 1>/dev/null 2>&1 || :  
== Cold: 	0:36.97 real,	17.44 user,	1.74 sys
= Command: bash -c cd test1200x10-xx-yy.clone; git annex drop . 1>/dev/null 2>&1 || :  
== Cold: 	0:37.15 real,	17.27 user,	1.62 sys
= Command: bash -c cd test1200x10-xx-yy-key.clone; git annex drop . 1>/dev/null 2>&1 || :  
== Cold: 	2:45.28 real,	83.39 user,	12.76 sys


= Command: rm -rf test1200x10-deref 
== Cold: 	0:00.19 real,	0.00 user,	0.17 sys
= Command: rm -rf test1200x10-bh1 
== Cold: 	0:04.69 real,	0.10 user,	2.04 sys
= Command: rm -rf test1200x10-bh1.clone 
== Cold: 	0:01.16 real,	0.02 user,	1.00 sys
= Command: rm -rf test1200x10-oh1 
== Cold: 	0:03.06 real,	0.08 user,	1.54 sys
= Command: rm -rf test1200x10-oh1.clone 
== Cold: 	0:01.37 real,	0.02 user,	1.17 sys
= Command: rm -rf test1200x10-oh1+bh1 
== Cold: 	0:04.81 real,	0.16 user,	2.14 sys
= Command: rm -rf test1200x10-oh1+bh1.clone 
== Cold: 	0:01.15 real,	0.01 user,	1.01 sys
= Command: rm -rf test1200x10-ohl 
== Cold: 	0:06.11 real,	0.19 user,	2.86 sys
= Command: rm -rf test1200x10-ohl.clone 
== Cold: 	0:01.38 real,	0.02 user,	1.24 sys
= Command: rm -rf test1200x10-x-xy 
== Cold: 	0:02.12 real,	0.04 user,	0.98 sys
= Command: rm -rf test1200x10-x-xy.clone 
== Cold: 	0:00.27 real,	0.00 user,	0.25 sys
= Command: rm -rf test1200x10-xx-y 
== Cold: 	0:02.12 real,	0.06 user,	0.96 sys
= Command: rm -rf test1200x10-xx-y.clone 
== Cold: 	0:00.24 real,	0.00 user,	0.22 sys
= Command: rm -rf test1200x10-xx-yy 
== Cold: 	0:02.33 real,	0.03 user,	1.02 sys
= Command: rm -rf test1200x10-xx-yy.clone 
== Cold: 	0:00.23 real,	0.00 user,	0.21 sys
= Command: rm -rf test1200x10-xx-yy-key 
== Cold: 	0:04.60 real,	0.13 user,	1.92 sys
= Command: rm -rf test1200x10-xx-yy-key.clone 
== Cold: 	0:02.60 real,	0.05 user,	2.17 sys

**Comment by joeyh** [comment]
Had a look at the details of removing the objectdir, and have put an implementation in a branch, removeobjectdir.

As noted above, freezing the hash directories can lead to races, and it would need a lock file. At worst, such a race would end with a permissions problem trying to write to a hash directory that had been frozen by another process. Not a big problem, and adding lock files wouldn't be too bad.

But there's also the problem that getKeysPresent walks .git/annex/objects to find  objects. However, object files are not the only files in there; there are also lock files, direct mode mapping files, etc. So, it currently looks for .git/annex/objects/xx/yy/$k/ and then checks if $k/$k exists. Without the $k directory, it can't look for $k/$k and so would need to find the files in .git/annex/objects/x/yy/ that are objects, and not locks etc. But, this cannot be determined from looking at the filename alone; "SHAE1-xxxx.lock" might be a lock file, or it might be an annexed file that happened to have the ".lock" extension.

So, seems that to implement this, the repo layout would need to put all the object files
in one set of hash directories, and the ancillary files elsewhere. Which becomes rather more complicated..

**Comment by joeyh** [comment]
Hmm, another reason it would need the lock files etc separated from the object files is, if a given filename can be either a lock file or an object file, than a repo could conceivably try to use the same filename for both purposes!

**Comment by yarikoptic** [comment]
I might be missing smth but why not to have something like a single .git/annex/objects/lock   which would list the file (or files if somehow done in batch mode + may be a PID of the process) which being modified (i.e.their yy directories given write permissions)?  then upon any operation there could even be a check if such file exists?

**Comment by joeyh** [comment]
Urk, getKeysPresent has more problems without per-object directories. If the lock/inodecache/mapping files are moved out of the way, then only object files will be in .git/annex/objects/, and it would make sense for getKeysPresent to just walk those. But, in direct mode, the object files won't be located there.

So, it would need to instead walk, I suppose, the inode cache files, when in direct mode without per-object directories. Except, it's possible for direct mode repos to have some objects stored in .git/annex/objects (ones that are not currently in the work tree). And, such objects are not going to have inode cache files.

So, it seems it would need to walk both the inode cache files, and whatever objects are in .git/annex/objects. But then it might find a key in both locations (perhaps due to a race or a stale inode cache file etc), and return a key twice. And since getKeysPresent streams lazily, it can't dedup without becoming a lot more memory and cpu expensive.

This is rapidly looking to be a hell of a mess. And it's not particularly clear that the resulting pile of hacks and extra work, extra locking, etc is going to be faster than the relatively simple and clean current implementation. I have to wonder how much a filesystem that is less slow than ZFS would improve the numbers.

**Comment by joeyh** [comment]
If http://git-annex.branchable.com/todo/cache_key_info/ were implemented, getKeysPresent could use its database of keys that are present in the annex.

So, this could be considered to be blocked by http://git-annex.branchable.com/design/caching_database/

**Comment by yarikoptic** [comment]
regarding other FSs: there also will be about double boost on such operations such as du (see e.g. above https://github.com/datalad/datalad/issues/32#issuecomment-70953059 or here is a full dump of different reports: http://www.onerussian.com/tmp/benchs.txt )

**Comment by yarikoptic** [comment]
just a note that if we thought that we would keep hitting filename length limits only under windows/FAT -- not quite so... backup to encfs'ed (fuse) directory started to fail:

```
ls: cannot access /home/xxx/_backups_/vault/c/xxxyyyyy/.git/annex/objects/mq/8m/URL--http&c%%audio.xxxyyyyy.net%audio%Prikliuchenia_Petrova_i_Vasechkina%plastinka2_-_K_nam_edet_huligan%&s28pesnia_iz_kf&s29_ne_verish.-33faff0bf383af1619db7ce9c5d9c04d/.URL--http&c%%audio.xxxyyyyy.net%audio%Prikliuchenia_Petrova_i_Vasechkina%plastinka2_-_K_nam_edet_huligan%&s28pesnia_iz_kf&s29_ne_verish.-33faff0bf383af1619db7ce9c5d9c04d.2BfYyl: File name too long
```
without intermediate key directory I had higher chance to succeed ;)

**Comment by yarikoptic** [comment]
For the reference -- https://git-annex.branchable.com/design/new_repo_versions/ describes new and future .git/annex store formats (to replace posted above link to v6)

**Comment by adswa** [comment]
I was wondering whether we can close this issue here. It has certainly had value and interesting discussions, but I don't really think it is worth keeping it open when the last activity here was 5.5 years ago, the general topic is more git-annex-y than datalad-y, and there were fixes and developments targeted at this issue here on git-annex side many years ago already. Okay to close?

**Comment by yarikoptic** [comment]
I agree -- we might reincarnate the topic but should be within git-annex discussion channels. IMHO issue persists.
